vllm/model_executor/guided_decoding/xgrammar_decoding.py

```diff
@@ -168,20 +186,22 @@ def from_guided_params(cls,
             else:
                 grammar_str = guided_params.grammar
             return cls(grammar_str=grammar_str,
-                       vocab_size=model_config.hf_config.vocab_size,
+                       vocab_size=tokenizer.vocab_size,
```

Comment:
@aarnphm there is a reason why we needed to reference the model's vocab size and not the tokenizers, correct?

Reply:
Cool, that was a shot, I knew that someone could clarify this.  I have already have a conflict on this part #11043 recently merged so I will revert this change. But would be nice to have documented the difference of these vocab sizes. @aarnphm do you have any reference for that?

Reply:
Thanks for check it out @mgoin! I really appreciate how fast you answered to this.

> Also the backend_str/vocab_type logic is a little confusing now.

Yeah, but the problem of this part is it was not correct. I reported a traceback on #11045.

```
from_huggingface(): incompatible function arguments. The following argument types are supported: (arg0: list[str], arg1: str, arg2: Optional[int], arg3: Optional[list[int]]) -> xgrammar.xgrammar_bindings.TokenizerInfo
```

First, `backend_str` is typed with str and was assigned with a Enum value (VocabType). And this function `from_huggingface` does not receive vocabType, and that's why it crashes using Mistral Tokenizer and possibly other types of tokenizers as well. So, for these cases we have call the contructor of `TokenizerInfo` and pass a vocabType (and no backend_str), which may be `RAW` for tokenizer like tikToken and `BYTE_FALLBACK` for mistral. I think maybe I can make it more explicit, first time I wrote a solution for that I added a bool field on `TokenizerData` like `is_from_huggingface`, but I decided to make it more concise and infer that based on the value of these values (which must be mutually exclusive). Of course I can add more comments to clarify and add asserts to prevent wrong behavior. 

What do you think?

Reply:
We need to reference tokenizer vocab size because of additional padding tokens.